Добавлена база H2 в проект

Natal1a-Chuklina opened the pull request.

Natal1a-Chuklina committed:
[r] FilmController.class, UserController.class - изменены типы возвращаемых значений и аргументов методов в связи с добавлением dto классов
[+] GenreController.class - добавлен контроллер для обработки запросов жанров
[+] MpaController.class - добавлен контроллер для обработки запросов рейтингов
[+] CreateFilmRequest.class, FilmResponse.class, UpdateFilmRequest.class - добавлены dto классы для формирования тела запросов и ответов для фильмов
[+] CreateUserRequest.class, UserResponse.class, UpdateUserRequest.class - добавлены dto классы для формирования тела запросов и ответов для пользователей
[+] FilmMapper.class, GenreMapper.class, MpaMapper.class, UserMapper.class - добавлены мапперы для перевода ответа сервера в объекты
[r, +] Film.class - изменена модель фильмов, добавлены новые поля
[+] Genre.class - добавлена модель для жанров фильмов
[+] Mpa.class - добавлена модель для рейтингов фильмов
[+] Status.enum - добавлено перечисления дли статусов дружбы
[r, +] User.class - изменена модель пользователя, добавлен статус дружбы
[r, +] FilmService.class - изменена логика работы сервиса из-за внедрения в проект бд, добавлены новые методы для обработки новых эндпоинтов
[r] UserService.class - изменена логика работы сервиса из-за внедрения в проект бд
[+] FilmDbStorage.class - добавлена реализация интерфейса FilmStorage для работы с бд
[+] UserDbStorage.class - добавлена реализация интерфейса UserStorage для работы с бд
[+] FilmStorage.interface, UserStorage.interface - добавлены новые методы в интерфейсы
[+] MpaStorage.interface, GenreStorage.interface - добавлены интерфейсы для взаимодействия с хранилищами рейтингов и жанров соответственно
[+] MpaDbStorage.class, GenreDbStorage.class - добавлены реализации интерфейсов MpaStorage и GenreStorage соответственно
[-] InMemoryFilmStorage.class, InMemoryUserStorage.class - классы удалены за ненадобностью из-за внедрения бд в проект
[+] Constants.class - добавлены новые константы
[+] FilmorateApplication.class - добавлен комментарий
application.properties - добавлена настройка бд
[+] data.sql - добавлен скрипт для заполнения бд стартовыми значениями
[+] schema.sql - добавлен скрипт для создания бд
[r] StartDateValidatorTest.class - изменены методы из-за внедрения dto
[-] FilmServiceTest.class, UserServiceTest.class - удалены тесты
[+] FilmorateApplicationTests - добавлено тестирование методов всех классов для работы с хранилищем
[+] clearDb.sql - добавлен скрипт для очистки бд
[+] pom.xml - добавлены зависимости в проект

Natal1a-Chuklina committed:
[r] FilmApplicationTests.class - изменен язык сообщений об ошибке на английский

z8q commented on src/main/java/ru/yandex/practicum/filmorate/dto/user/UserResponse.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+@Getter
+@Setter
+public class UserResponse {
```
Использовать dto это хорошо.
Но реализовано не совсем правильно. Можно посоветоваться с наставником, но рекомендую все таки использовать либо одну сущность, либо одну dto.
Если какие-то поля не используются в сценариях. Например при create нам принимать id не нужно,
Мы можем просто не обрабатывать это поле. Задавать id просто перед сохранением сущности.

Natal1a-Chuklina replied:
Переделала.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    @Override
+    public int add(CreateFilmRequest film) {
```
Отправляю POST запрос такого вида:
{
	"name": "Die Hard",
	"description": "Action movie with Br-s Wil-s",
	"releaseDate": "1988-07-22",
	"duration": 1234567,
	"rating": 100,
	"mpa": {
		"id": 5,
		"name":"NC-17",
		"description": "17+"
	},
	"genres": [ `в жанрах не указал id, чтобы появлялась ошибка`
		{
			"name": "action"
		},
		{
			"name": "drama"
		},
		{
			"name": "comedy"
		}
	]
}

Приходит такая ошибка:
{
	"errorMessage": "PreparedStatementCallback; SQL [INSERT INTO film_genre (film_id, genre_id) VALUES (?, ?);]; Referential integrity constraint violation: \"FK_GENRE_ID: PUBLIC.FILM_GENRE FOREIGN KEY(GENRE_ID) REFERENCES PUBLIC.GENRES(ID) (0)\"; SQL statement:\nINSERT INTO film_genre (film_id, genre_id) VALUES (?, ?) [23506-214]; nested exception is org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: Referential integrity constraint violation: \"FK_GENRE_ID: PUBLIC.FILM_GENRE FOREIGN KEY(GENRE_ID) REFERENCES PUBLIC.GENRES(ID) (0)\"; SQL statement:\nINSERT INTO film_genre (film_id, genre_id) VALUES (?, ?) [23506-214]"
}

Не очень хорошо показывать пользователю внутренние sql запросы. Ошибка должна быть более понятная.
Плюс фильм в итоге сохраняется не смотря на ошибку.
Было бы правильно не сохранять объект, если возникает ошибка.
Также программа допускает добавление фильмов с одинаковыми полями, можно ограничить.

Запросы можно отправлять через postman или insomnia.
Смотреть результаты сохранения либо в базе через идею, либо через консоль http://localhost:8080/h2-console.
Или можно просто get запросами смотреть, сохранился объект после ошибки или нет

Natal1a-Chuklina replied:
Тут я увидела сразу несколько проблем и их исправила:
1) При таком запросе у меня должна была возвращаться ошибка валидации, но так как я допустила ошибку этого не происходило, ошибку исправила, все работает.
2) В случае неизвестных ошибок пользователю выводилось сообщение ошибки (e.getMessage()). В данной ситуации это сообщение содержало SQL запрос, чтобы в дальнейшем не вывести случайно лишнюю информацию я заменила вывод сообщения ошибки на вывод заранее заданного мной сообщения.

Проблему добавления фильмов с одинаковыми полями решила. Теперь если у фильмов совпадают имя, описание, дата релиза и длительность, то фильм не добавляется.

И постманом, и консолью я активно пользуюсь. В консоли было удобно составлять запросы для проекта, так как сразу было видно, получаю ли я всю нужную информацию или нет, так же были видны дублирования названий колонок, которые в дальнейшем моли привести к ошибке.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -20,25 +22,25 @@ public UserController(UserService userService) {
 
     @GetMapping
-    public Collection<User> getAll() {
+    public Collection<UserResponse> getAll() {
```
Когда запрашиваем пользователей, можно было бы показывать поле с друзьями тоже

Natal1a-Chuklina replied:
Добавила.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,207 @@
+    @Override
+    public Collection<Film> getBestFilms(int count) {
+        String sql =
+                "SELECT f.id, " +
```
Sql запросы по всему коду красиво сделаны

Natal1a-Chuklina committed:
[+, r] ErrorHandler.class - добавлена обработка нового исключения, в случае неизвестной ошибки вывод сообщения об ошибке заменен на универсальное сообщение
[r] FilmController.class - класс изменен в связи с удалением Dto классов
[r] UserController.class - класс изменен в связи с удалением Dto классов
[-] CreateFilmRequest.class, FilmResponse.class, UpdateFilmRequest.class, CreateUserRequest.class,  UpdateUserRequest.class, UserResponse.class - удалены классы
[+] FilmMapper.class - добалено заполнение поля лайков информацией из бд
[+] Film.class, User.class - дополнены аннотациями, конструкторами
[r] Genre.class, Mpa.class - исправлена ошибка в аннотациях
[r, +] FilmService.class - класс изменнен в связи с удалением dto классов, добавлены запросы в бд после создания/изменения фильма для получения актуальной информации
[r, +] UserService.class - класс изменнен в связи с удалением dto классов, добавлены запросы в бд после создания/изменения пользователя для получения актуальной информации
[r] FilmStorage.interface, UserStorage.interface, UserDbStorage.class, FilmDbStorage.class - изменены аргументы методов
[+] FilmDbStorage.class - дополнены sql запросы для получения лайков фильма
[+] Constants.class - добавлены константы
[-] FilmorateApplication.class - удален комментарий
[+] schema.sql - добавлено ограничение, чтобы нельзя было создать идентичные фильмы несколько раз
[r] StartDateValidatorTest.class, FilmorateApplicationTests.class - тесты изменены из-за удаления Dto классов
